### DISCUSSION TITLE
How do I set the 3D coordinates of the virtual camera when rendering point clouds?

### DISCUSSION THREAD

**Discussion by Zhouwudexiazhou** [isl-org/Open3D#7031] [General]
I want to simulate the situation of people watching the volume video with the head display, because the image rendered by the point cloud is different from the distance. However, in oped3d, there is only the look at function of the viewing point, but there is no function to set the camera's 3d space coordinates. I want to ask how can I set the camera's desired 3d coordinates to render the point cloud picture?

**Comment by cwreynolds** [comment]
Hi. Which visualizer are you using? There is a newer `open3d::visualization::visualizer::O3DVisualizer` and the older “legacy” `open3d::visualization::Visualizer`. (Also which language? Most people use python. I'm using c++ names here.)

In the first case you should be able to use `O3DVisualizer::SetupCamera(fov, center, eye, up)`

In the legacy case, I had exactly the same confusion: why can't I directly set the “eye” point of `ViewControl`?! I recently described my work-around here: #7023 (“look at” add-on to legacy Visualizer/ViewControl).

**Reply by Zhouwudexiazhou** [reply]
I use python and the older “legacy”，I will try what you said tomorrow and reply you

**Reply by Zhouwudexiazhou** [reply]
The methods you mentioned are based on c++, are there any python implementation methods?

**Reply by cwreynolds** [reply]
I think the answer is yes, but I have not used Open3D from Python for a year or so. So I cannot quickly prototype a version of my work-around in Python.

I did find the Python API doc for [look_at()](open3d.visualization.rendering.Camera.look_at). See also the legacy Visualizer method [get_view_control()](https://www.open3d.org/docs/latest/python_api/open3d.visualization.Visualizer.html#open3d.visualization.Visualizer.get_view_control), the class [ViewControl](https://www.open3d.org/docs/latest/python_api/open3d.visualization.ViewControl.html#open3d.visualization.ViewControl), and its methods [convert_from_pinhole_camera_parameters()](https://www.open3d.org/docs/latest/python_api/open3d.visualization.ViewControl.html#open3d.visualization.ViewControl.convert_from_pinhole_camera_parameters) / [convert_to_pinhole_camera_parameters()](https://www.open3d.org/docs/latest/python_api/open3d.visualization.ViewControl.html#open3d.visualization.ViewControl.convert_to_pinhole_camera_parameters).

Perhaps @ssheorey can suggest someone with better experience to help you translate my c++ code to Python?

**Reply by cwreynolds** [reply]
Maybe the answer is **no**. I tried to translate my code to Python. That required making a new environment for Open3D 0.18.0 since the last time I used Open3D from Python the current release was 0.17.0, which had a bug preventing modification of `ViewControl`.

But as near as I can tell, there is no direct Python analog to the c++ function `open3d::visualization::gl_util::LookAt()` which my code is based on. The symbol “gl_util” does not appear in the Open3D Python doc at all. It would be possible to write such a function, but I think that is probably going too far in the wrong direction. 

You might be better off to look into the newer `O3DVisualizer` class which has a function providing look-at-up functionality: [setup_camera()](https://www.open3d.org/docs/release/python_api/open3d.visualization.O3DVisualizer.html#open3d.visualization.O3DVisualizer.setup_camera)

**Reply by Zhouwudexiazhou** [reply]
I have tested O3DVisualizer class in python. It work!!! Thank you bro,Thank you!!!

**Reply by Zhouwudexiazhou** [reply]
> 也许答案是**否定的**。我尝试将我的代码翻译成 Python。这需要为 Open3D 0.18.0 创建一个新环境，因为上次我从 Python 使用 Open3D 时，当前版本是 0.17.0，它有一个错误阻止修改`ViewControl`。
> 
> 但据我所知，没有与`open3d::visualization::gl_util::LookAt()`我的代码所基于的 C++ 函数直接对应的 Python 函数。符号“gl_util”根本没有出现在 Open3D Python 文档中。可以编写这样的函数，但我认为这可能走得太远了。
> 
> 你最好研究一下较新的`O3DVisualizer`类，它有一个提供查找功能的函数：[setup_camera()](https://www.open3d.org/docs/release/python_api/open3d.visualization.O3DVisualizer.html#open3d.visualization.O3DVisualizer.setup_camera)

hey,bro.I read the C++ version of O3DVisualizer class code.When i change the fov in setupcamera.Why the photo change together?It do not change!! the photo1 with the fov value in 10 degree and photo2 with fov value in 60 degree are the same!!
The C++ code is:
![image](https://github.com/user-attachments/assets/c6f33eea-6f23-4a25-a2a4-718264bce90c)
It does not  use the fov!!

**Reply by cwreynolds** [reply]
Good catch @Zhouwudexiazhou! I added a comment on your issue #7033 pointing back to your comment just above.
